crates/red_knot_python_semantic/src/suppression.rs

```diff
@@ -1,50 +1,100 @@
-use salsa;
+use std::cmp::Ordering;
 
-use ruff_db::{files::File, parsed::comment_ranges, source::source_text};
+use ruff_python_parser::TokenKind;
+use ruff_source_file::LineRanges;
+use ruff_text_size::{Ranged, TextRange, TextSize};
+
+use ruff_db::{files::File, parsed::parsed_module, source::source_text};
 use ruff_index::{newtype_index, IndexVec};
 
 use crate::{lint::LintId, Db};
 
 #[salsa::tracked(return_ref)]
-pub(crate) fn suppressions(db: &dyn Db, file: File) -> IndexVec<SuppressionIndex, Suppression> {
-    let comments = comment_ranges(db.upcast(), file);
+pub(crate) fn suppressions(db: &dyn Db, file: File) -> Suppressions {
     let source = source_text(db.upcast(), file);
+    let parsed = parsed_module(db.upcast(), file);
 
     let mut suppressions = IndexVec::default();
+    let mut line_start = source.bom_start_offset();
+
+    for token in parsed.tokens() {
+        match token.kind() {
+            TokenKind::Comment => {
+                let text = &source[token.range()];
+
+                let suppressed_range = TextRange::new(line_start, token.end());
 
-    for range in comments {
-        let text = &source[range];
-
-        if text.starts_with("# type: ignore") {
-            suppressions.push(Suppression {
-                target: None,
-                kind: SuppressionKind::TypeIgnore,
-            });
-        } else if text.starts_with("# knot: ignore") {
-            suppressions.push(Suppression {
-                target: None,
-                kind: SuppressionKind::KnotIgnore,
-            });
+                if text.strip_prefix("# type: ignore").is_some_and(|suffix| {
+                    suffix.is_empty() || suffix.starts_with(char::is_whitespace)
+                }) {
```

Comment:
does this mean we ignore `type: ignore` comments that use mypy error codes? That's different to e.g. what pyright does. If pyright sees a `type: ignore[misc]` comment, they just interpret it as a bare `type: ignore` comment that suppresses all typing violations (because neither tool supports using `type: ignore` comments with error codes).

This has proved to be very useful in e.g. typeshed, where we often have `type: ignore` comments that use mypy error codes. Pyright just treats them as blanket `type: ignore` comments in our CI at typeshed, but mypy understands them as only applying to specific mypy error codes.

Pytype is the third type checker that we run in CI at typeshed, but I believe they do not support `type: ignore` comments at all, so anything we put in the `type: ignore` comment is sort-of irrelevant from pytype's perspective

Reply:
Oh nice find. Let me add a test for this and also take a look at pyrights regex

It's nit entirely clear to me yet whether we should support type: ignore comments with codes or do what pyright does. I can see how ignoring them is useful in a situation like typeshed but I could see why users would want to use them

Reply:
I don't think we should support `type: ignore` comments with our own codes. This will cause us a practical problem right out of the gate, because typeshed uses mypy codes in `type: ignore` comments, and our codes don't match mypy codes. If we did this, we'd also have to implement compatibility code aliases for mypy codes right away, which is a complexity we probably want to avoid?

For better or worse, mypy already claimed the "code namespace" for `type: ignore`; given the typeshed situation I think our best option is to do what pyright does: accept and ignore codes with `type: ignore`. And use our own codes with our own form of suppression comment.

Reply:
I agree that in terms of the principle it should not be, and I totally get that it's frustrating to have our design space constrained here. But I agree with Carl that it will probably cause too many compatibility problems if we reject or ignore all `type: ignore` comments that use mypy error codes. That will make it quite painful for users to migrate from mypy to red-knot, or for users to run both tools on one repo. Obviously we hope that most users won't need to run multiple type checkers on any one repo, but some users such as typeshed will *have* to.

Reply:
That makes sense. I'm also not reasoning about it nor feeling frustrated. I just think that users will ask for it and we may want to add support for it in some way or another. It could just be an option.

Ignoring codes also isn't necessarily enough to run mypy and red knot side by side. At least not if red knot warns about unused suppressions. I don't think we have to solve this now and not supporting codes seems a good start. But we may need a little more but let's learn from users and testing on real projects

Reply:
I feel quite strongly that we should avoid creating migration barriers unnecessarily, and mypy is currently the most popular type checker, so I do think we will sometimes have to make decisions based on what it does.

Supporting our own codes in `type: ignore`, without mypy compat aliases, unnecessarily creates a migration barrier from mypy, so I think it would be shooting ourselves in the foot to do this without very strong reasons.